xcm/src/v1/multilocation.rs

```diff
@@ -329,18 +329,40 @@ impl MultiLocation {
 	///
 	/// Does not modify `self` in case of overflow.
 	pub fn reanchor(&mut self, target: &MultiLocation, ancestry: &MultiLocation) -> Result<(), ()> {
-		// TODO: https://github.com/paritytech/polkadot/issues/4489 Optimize this.
-
-		// 1. Use our `ancestry` to figure out how the `target` would address us.
-		let inverted_target = ancestry.inverted(target)?;
-
-		// 2. Prepend `inverted_target` to `self` to get self's location from the perspective of
-		// `target`.
-		self.prepend_with(inverted_target).map_err(|_| ())?;
+		if target.parents == self.parents {
+			// Find the first junction that is different as we travel along the paths from ancestry to self and target
+			let mut first_different = None;
+			for i in 0..target.interior.len() {
+				if target.interior.at(i) != self.interior.at(i) {
+					first_different = Some(i);
+					break
+				}
+			}
 
-		// 3. Given that we know some of `target` ancestry, ensure that any parents in `self` are
-		// strictly needed.
-		self.simplify(target.interior());
+			if let Some(i) = first_different {
+				// In the case where the self and target paths split, remove just the shared path
+				self.parents = (target.interior.len() - i) as u8;
+				for _ in 0..i {
+					self.take_first_interior();
+				}
+			} else {
+				// In the case where the paths are the same all the way to the target,
+				// remove all the junctions on the shared path
+				for _ in 0..target.interior.len() {
+					self.parents = 0;
+					self.take_first_interior();
+				}
+			}
+		} else if target.parents > self.parents {
+			// Handle when the target has more parents so we have to go through ancestry
+			self.parents = target.len() as u8 - 1;
+			self.interior
+				.push_front(ancestry.interior.first().unwrap_or(&Junction::OnlyChild).clone())
+				.map_err(|_| ())?;
+		} else if target.parents < self.parents {
+			// Handle when the id has more parents so we must go all the way up the target path
+			self.parents = self.parents + target.interior().len() as u8;
```

Comment:
If `self`/`id` has more parents, then it doesn't sound like it's simply an addition here, I'd expect to see something like `(self.parents - target.parents) + target.interior().len() as u8`. The `self.parents - target.parents` part is the path used by the `target` to reach the part of `ancestry` in which `self`/`id` actually branches off from and starts concatenating interior junctions.